module/os/linux/spl/spl-proc.c

```diff
@@ -648,21 +651,36 @@ static struct ctl_table spl_dir[] = {
 
 static struct ctl_table spl_root[] = {
 	{
-	.procname = "kernel",
-	.mode = 0555,
-	.child = spl_dir,
+		.procname	= "kernel",
+		.mode		= 0555,
+		.child		= spl_dir,
 	},
 	{}
 };
+#endif
 
 int
 spl_proc_init(void)
 {
 	int rc = 0;
 
-	spl_header = register_sysctl_table(spl_root);
+#ifdef HAVE_REGISTER_SYSCTL_TABLE
+	spl_header = register_sysctl_table(spl_table);
 	if (spl_header == NULL)
 		return (-EUNATCH);
+#else
+	spl_header = register_sysctl("kernel/spl", spl_table);
+	if (spl_header == NULL)
+		return (-EUNATCH);
```

Comment:
`unregister_sysctl_table()` below is expecting the root of the hierarchy to tear everything down.  With the restructuring it looks like we're not cleaning up properly on error or during `spl_proc_fini()`.

Reply:
In the new version, I preserve `spl_header` as containing the pointer of the root node, and just test the return values for `NULL` in the child node error cases, setting `rc` appropriately and doing a `goto out` as is convention later on in this function. The `unregister_sysctl_table()` function in 6.5 still does a descendant walk & teardown, so as long as `spl_header` is pointing at the root node in both implementations, it seems like it should be able to remain identical in both the new and old kernel cases.